storage/testonly/admin_storage_tester.go

```diff
@@ -0,0 +1,309 @@
+		{newTrees: 3}, // total 4
+	}
+
+	s := tester.NewAdminStorage()
+
+	wantTrees := []*trillian.Tree{}
+	for i, test := range tests {
+		// Setup new trees (as specified)
+		if test.newTrees > 0 {
+			before := len(wantTrees)
+			ctx := context.Background()
+			tx, err := s.Begin(ctx)
+			if err != nil {
+				t.Fatalf("%v: Begin() = %v, want = nil", i, err)
+			}
+			defer tx.Close()
+			for j := 0; j < test.newTrees; j++ {
+				tree, err := tx.CreateTree(ctx, LogTree)
+				if err != nil {
+					t.Fatalf("%v: CreateTree() = (_, %v), want = (_, nil)", i, err)
+				}
+				wantTrees = append(wantTrees, tree)
+			}
+			if err := tx.Commit(); err != nil {
+				t.Fatalf("%v: Commit() = %v, want = nil", i, err)
+			}
+			if got := len(wantTrees) - before; got != test.newTrees {
+				t.Fatalf("got %v new trees, want = %v", got, test.newTrees)
+			}
+		}
+
+		ctx := context.Background()
+		tx, err := s.Snapshot(ctx)
+		if err != nil {
+			t.Fatalf("%v: Snapshot() = %v, want = nil", i, err)
+		}
+		runListTreeIDsTest(ctx, t, i, tx, wantTrees)
+		runListTreesTest(ctx, t, i, tx, wantTrees)
+		if err := tx.Commit(); err != nil {
+			t.Errorf("%v: Commit() = %v, want = nil", i, err)
+		}
+	}
+}
+
+func runListTreeIDsTest(ctx context.Context, t *testing.T, i int, tx storage.ReadOnlyAdminTX, wantTrees []*trillian.Tree) {
+	ids, err := tx.ListTreeIDs(ctx)
+	if err != nil {
+		t.Errorf("%v: ListTreeIDs() = (_, %v), want = (_, nil)", i, err)
+		return
+	}
+	if got, want := len(ids), len(wantTrees); got != want {
+		t.Errorf("%v: got len(ids) = %v, want = %v", i, got, want)
+		return
+	}
+	wantIDs := []int64{}
+	for _, tree := range wantTrees {
+		wantIDs = append(wantIDs, tree.TreeId)
+	}
+	if got, want := toIntMap(ids), toIntMap(wantIDs); !reflect.DeepEqual(got, want) {
+		t.Errorf("%v: ListTreeIDs = (%v, _), want = (%v, _)", i, ids, wantIDs)
+		return
+	}
+}
+
+func runListTreesTest(ctx context.Context, t *testing.T, i int, tx storage.ReadOnlyAdminTX, wantTrees []*trillian.Tree) {
+	trees, err := tx.ListTrees(ctx)
+	if err != nil {
+		t.Errorf("%v: ListTrees() = (_, %v), want = (_, nil)", i, err)
+		return
+	}
+	if got, want := len(trees), len(wantTrees); got != want {
+		t.Errorf("%v: got len(ids) = %v, want = %v", i, got, want)
+		return
+	}
+	if got, want := toTreeMap(trees), toTreeMap(wantTrees); !reflect.DeepEqual(got, want) {
+		t.Errorf("%v: ListTrees = (%v, _), want = (%v, _)", i, trees, wantTrees)
+		return
+	}
+}
+
+func toIntMap(values []int64) map[int64]bool {
+	m := make(map[int64]bool)
+	for _, v := range values {
+		m[v] = true
+	}
+	return m
+}
+
+func toTreeMap(values []*trillian.Tree) map[int64]*trillian.Tree {
+	m := make(map[int64]*trillian.Tree)
+	for _, v := range values {
+		m[v.TreeId] = v
+	}
+	return m
+}
+
+// TestAdminTXClose verifies the behavior of Close() with and without explicit Commit() / Rollback() calls.
+func (tester *AdminStorageTester) TestAdminTXClose(t *testing.T) {
+	tests := []struct {
+		commit       bool
```

Comment:
worth adding a "desc string" member to the tests and adding it to the log msgs below (rather than index)?